### ISSUE TITLE
Feature Request: Mongo(DB)Dsn type

### ISSUE THREAD

**Issue by StephenBrown2** [samuelcolvin/pydantic#2117]
### Checks

* [x] I added a descriptive title to this issue
* [x] I have searched (google, github) for similar issues and couldn't find anything
* [x] I have read and followed [the docs](https://pydantic-docs.helpmanual.io/) and still think this feature/change is needed
* [x] After submitting this, I commit to one of:
    * Look through open issues and helped at least one other person
    * Hit the "watch" button on this repo to receive notifications and I commit to help at least 2 people that ask questions in the future
    * Implement a Pull Request for a confirmed bug

# Feature Request

Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
             pydantic version: 1.7.2
            pydantic compiled: True
                 install path: /Users/step7212/.pyenv/versions/3.9.0/envs/chas3/lib/python3.9/site-packages/pydantic
               python version: 3.9.0 (default, Oct  5 2020, 14:52:54)  [Clang 11.0.0 (clang-1100.0.33.17)]
                     platform: macOS-10.14.6-x86_64-i386-64bit
     optional deps. installed: ['typing-extensions', 'devtools']
```

```py
import pydantic

class MyModel(pydantic.BaseModel):
   mongodb: pydantic.MongodbDsn

MyModel(mongodb="mongodb://mongodb1.example.com:27317,mongodb2.example.com:27017/?replicaSet=mySet&authSource=authDB")
```

The MongoDB URI scheme is defined here: https://www.iana.org/assignments/uri-schemes/prov/mongodb
as: `mongodb://[username:password@]host1[:port1][,...hostN[:portN]][/[defaultauthdb][?options]]`
some examples of which can be found in the Reference: https://docs.mongodb.com/manual/reference/connection-string/
```
mongodb://mongodb0.example.com:27017
mongodb://myDBReader:D1fficultP%40ssw0rd@mongodb0.example.com:27017/?authSource=admin
mongodb://mongodb1.example.com:27317,mongodb2.example.com:27017/?replicaSet=mySet&authSource=authDB
mongodb://db0.example.com,db1.example.com,db2.example.com/?replicaSet=myRepl&tls=true
```

The sticking point is, I think, the potential for multiple hosts, i.e. domain:port combinations. The current regex only grabs a single host, separate from an optional port. If we were to modify it to group the host and port sections with a repeating operator, Python's `re` module can't capture multiple groups, while the third-party `regex` package can. However, that would add a new dependency to pydantic, which we _probably_ don't want to do.

Another option is to group the whole thing, and split and parse it manually, or just scrap the pattern and write a new one, since the fragment and other items aren't used, and the `query` string is called `options`, and the `path` is used for the `database` to auth to.

A couple other notes lead me to the latter option, namely:

> If the username or password includes the following characters:
> ```
> : / ? # [ ] @
> ```
> those characters must be converted using percent encoding.

and some backwards compatibility:

> **Semi-colon separator for connection string arguments**
> 
> To provide backwards compatibility, drivers currently accept semi-colons (i.e. ;) as option separators.

While I would love to re-use as much functionality from `AnyUrl` as already exists, the quirks of MongoDB connection strings seems to outweigh the similarities it has.

If anyone wants to take a crack at implementing a `MongodbDsn` type, or has a good idea on what to start on for the regex, that'd be great. I just want to put this out there to gauge interest and difficulty of implementation given the above quirks and potentially others that I may not have noticed as of yet.

**Comment by art049** [comment]
Including as well the [DNS Seed List Connection Format](https://docs.mongodb.com/manual/reference/connection-string/#dns-seed-list-connection-format) for connection string like: `mongodb+srv://server.example.com/` could be handy as well.

**Comment by StephenBrown2** [comment]
> Including as well the [DNS Seed List Connection Format](https://docs.mongodb.com/manual/reference/connection-string/#dns-seed-list-connection-format) for connection string like: `mongodb+srv://server.example.com/` could be handy as well.

I agree, though I think that it would be only a matter of allowing `mongodb+srv` as the scheme; the application itself would/should handle the SRV record fetching.

**Comment by art049** [comment]
> I agree, though I think that it would be only a matter of allowing mongodb+srv as the scheme; the application itself would/should handle the SRV record fetching.

Yes, totally. I meant mainly this to prevent runtime issues. An application supporting the `mongodb` scheme would not necessarily support the `mongodb+srv` scheme (because of the record fetching that requires an extra package comparing to regular connection strings).
So if you want only traditional connection string, using `MongodbDsn` would work but it would reject the DNS style connection strings.
And another one like `MongoSRVDsn`  would allow both `mongodb+srv` and `mongodb` schemes.

**Comment by art049** [comment]
Would it be interesting to extract the options as well ?
`mongodb://mongodb1.example.com:27317,mongodb2.example.com:27017/mydatabase?replicaSet=mySet&authSource=authDB `
Could give the following attributes to the field:
- hosts: `("mongodb1.example.com:27317", "mongodb2.example.com:27017")` (should the port/host be separated as well)
- authentication_database: `"authDB"` (takes precedence over `"mydatabase"` if both are specified)
- replica_set:  `"mySet"`

As there is quite a lot of other parameters I'm not quite sure it's still in the scope of Pydantic.
Maybe even the `authentication_database` and `replica_set` I mentionned before should not be generated and only the query string should be kept for them.

**Comment by StephenBrown2** [comment]
> Would it be interesting to extract the options as well ?

I think that the options should be separated out, but they can be their own dict. It'd be nifty if there was a submodel with all the possible options, but that is subject to change with mongo release versions, so I think you're right that it's out of scope and keeping it more generic is better.

For `mongodb://mongodb1.example.com:27317,mongodb2.example.com:27017/mydatabase?replicaSet=mySet&authSource=authDB`

the hosts can be their own types with `host` and `port` attributes, which defaults to `27017` I would imagine something like the following, using rules as given in https://docs.mongodb.com/manual/reference/connection-string/ for things like the auth_source :

```
scheme: mongodb
username: None
password: None
hosts:
  - host: mongodb1.example.com
    port: 27317
  - host: mongodb2.example.com
    port: 27317
default_auth_db: mydatabase
auth_source: authDB  # Would be "mydatabase" if authSource not given in options, and "admin" if neither
options:
  replica_set: mySet
  auth_source: authDB
```

All the options can be aliased with this alias generator:
```py
def snake_to_pascal(string: str) -> str:
    split = string.split("_")
    return split[0] + "".join(word.capitalize() for word in string.split("_")[1:])
```
(Or I suppose the reverse would need to occur)

**Comment by StephenBrown2** [comment]
Per https://github.com/samuelcolvin/pydantic/pull/2118#issuecomment-741297615 I think that I may have gotten too focused on "nice-to-haves", when this should just be a simple validation rather than parsing.

This should still be similar to the URL parsing, but the only differences would be an additional `hosts` attribute, which should be a submodel including a hostname (default to `localhost`) and a port (default to `27017`), also potentially including `host_type`.

`allowed_schemes` would be `mongodb`

Attributes should be (based on current `AnyUrl` class):

scheme: str
user: str
password: str
~~host~~
~~tld~~ (Unsure if this is needed/desired, as it matters not for a mongo dsn, so could be removed/ignored)
hosts: List[Host]{host: str, port: int, type: str}
~~host_type~~
~~port~~
path: str
query: str
options: Dict (parsed `query` with `snake_cased` key aliases, _**optional!**_)
~~fragment~~ (unused in mongo DSNs)

**Comment by duarte-pompeu** [comment]
It seems this was partially implemented, but doesn't support replicas: https://github.com/pydantic/pydantic/pull/3230.

I was fiddling around and found that:

- `MongoDsn` extends `AnyUrl`
- `PostgresDsn` extends `MultiHostDsn`. 

Looking at the tests for PostgresDSN, the URLs seem very similar to MongoDB URLs for replicas: perhaps we could also inherit from it?

References:

https://github.com/pydantic/pydantic/blob/1a5b56e1b2020fd74800064d038becf614d64a61/pydantic/networks.py#L531-L539

https://github.com/pydantic/pydantic/blob/1a5b56e1b2020fd74800064d038becf614d64a61/pydantic/networks.py#L487-L500

https://github.com/pydantic/pydantic/blob/1a5b56e1b2020fd74800064d038becf614d64a61/tests/test_networks.py#L424-L437

**Comment by samuelcolvin** [comment]
This is actually parially fixed in #3349, but it was only implemented for postgresql, not mongo.

More exciting, I've just implemented muti-host URLs in rust as part of pydantic-core, see https://github.com/pydantic/pydantic-core/pull/325.

I'd really appreciate it anyone here who has a "dig in the fight" of complex DSN parsing could take a look at https://github.com/pydantic/pydantic-core/pull/325 and in particular contribute anymore difficult cases to check.
